src/btree/bt_curnext.c

```diff
@@ -603,18 +603,26 @@ __wt_btcur_next(WT_CURSOR_BTREE *cbt, bool truncating)
 	 */
 	F_CLR(cbt, WT_CBT_RETRY_PREV);
 	if (F_ISSET(cbt, WT_CBT_RETRY_NEXT)) {
-		WT_RET(__wt_cursor_valid(cbt, &upd, &valid));
+		WT_ERR(__wt_cursor_valid(cbt, &upd, &valid));
+		WT_ASSERT(session, valid == true);
+
+		/* The update that returned prepared conflict is now visible. */
 		F_CLR(cbt, WT_CBT_RETRY_NEXT);
-		if (valid) {
-			/*
-			 * If the update, which returned prepared conflict is
-			 * visible, return the value.
-			 */
-			return (__cursor_kv_return(session, cbt, upd));
-		}
+
+		/*
+		 * The underlying key-return function uses a comparison value
+		 * of 0 to indicate the search function has pre-built the key
+		 * we want to return. That's not the case, don't take that path.
+		 */
+		cbt->compare = 1;
+		WT_ERR(__cursor_kv_return(session, cbt, upd));
```

Comment:
I think this can handle even when upd == NULL ?
If the prepared_updated we were pointing to is committed and if it is a delete operation, then we should continue and get the next value. My understanding is this is already handled as part of __cursor_kv_return.

Reply:
That's one of the reasons I flagged this change for you. I wasn't sure if a prepare-conflict followed by a committed delete would return `WT_NOTFOUND` or `0 with upd=NULL, valid=false`. If it's the latter, then I think I agree, we should fall out of that clause into a normal iteration.

I don't understand what you mean by "handled as part of __cursor_kv_return". What specific code are you thinking of that handles this?